commit a0f6d6caef4033aa9c3e2ea2ceae256c4347a419
Author: Amitkumar Karwar <akarwar@marvell.com>
Date:   Fri Feb 24 21:36:05 2012 -0800

    mwifiex: handle auto authentication mode correctly
    
    When authentication type is configured to NL80211_AUTHTYPE_AUTOMATIC,
    driver tries to connect using open mode. The association is failed
    if AP is configured in shared mode.
    
    This patch adds code to try association using shared mode as well if
    open mode association fails.
    
    Now since we returned exact error code in association response handler
    (instead of -1), corresponding changes are done in
    mwifiex_process_cmdresp().
    
    Signed-off-by: Amitkumar Karwar <akarwar@marvell.com>
    Signed-off-by: Bing Zhao <bzhao@marvell.com>
    Signed-off-by: John W. Linville <linville@tuxdriver.com>

diff --git a/drivers/net/wireless/mwifiex/sta_ioctl.c b/drivers/net/wireless/mwifiex/sta_ioctl.c
index 866026ecca44..e10161f5894c 100644
--- a/drivers/net/wireless/mwifiex/sta_ioctl.c
+++ b/drivers/net/wireless/mwifiex/sta_ioctl.c
@@ -249,6 +249,17 @@ int mwifiex_bss_start(struct mwifiex_private *priv, struct cfg80211_bss *bss,
 		 * application retrieval */
 		priv->assoc_rsp_size = 0;
 		ret = mwifiex_associate(priv, bss_desc);
+
+		/* If auth type is auto and association fails using open mode,
+		 * try to connect using shared mode */
+		if (ret == WLAN_STATUS_NOT_SUPPORTED_AUTH_ALG &&
+		    priv->sec_info.is_authtype_auto &&
+		    priv->sec_info.wep_enabled) {
+			priv->sec_info.authentication_mode =
+						NL80211_AUTHTYPE_SHARED_KEY;
+			ret = mwifiex_associate(priv, bss_desc);
+		}
+
 		if (bss)
 			cfg80211_put_bss(bss);
 	} else {